keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1230 @@
+but there are many details to consider.
+
+### Packaging
+
+<<[UNRESOLVED where it lives]>>
+
+The simplest thing to do is to write a new implementation of k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most obviously reusable place for it.
+
+A more ecosystem-friendly solution is to write a new function or package under  sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.
+
+<<[/UNRESOLVED]>>
+
+### Tooling
+
+k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and ensures they are formatted “correctly”. The only meaning of “correct” is “block style”, and it corrupts KYAML input today (aside: it also does not handle multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML output, and set the default to KYAML (or to change the parent script to always specify KYAML).
+
+This tooling will preserve comments.
+
+#### How?
+
+Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While KYAML is an output format, our tooling goals require that we get access to a YAML AST, which may require some dirty tricks.  If we ever change YAML libraries (e.g. [https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may need different dirty tricks.
+
+Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST, which we can render as KYAML.  Neither, however, offers a way to encode an arbitrary Go object as the equivalent YAML AST.  To do that we may have to encode the object to plain YAML, then decode it to AST, then render it as KYAML.  That extra conversion is ugly, but this is not generally performance-sensitive code and is not run on the server.
+
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an API which appears to do what we need in both directions.
+
+### Header
+
+At the top of each KYAML document (or multi-document) will be the following header:
+
+```
+#!yaml
+---
+```
+
+This disambiguates it from JSON, since both KYAML and JSON begin with an opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML without the header, but it is needed for versions before that.
+
+### Per-type rendering
+
+Integers and floats are rendered as their natural numeric representation.
+
+Simple strings are rendered as double-quoted and escaped. Multi-line strings require special consideration, see below.
+
+Booleans are rendered as `true` or `false`.
+
+Structs are rendered with curly-braces. Field names are not quoted unless they are not obviously strings (which includes prefixed label keys) or are one of the known type-ambiguous words (e.g. `no`).  Field values are rendered as per their type. Embedded structs will be rendered as inline fields.  Fields which have a `json` struct tag will be rendered according to that tag:
+  * the tag’s field name will be used
```

Comment:
Doesn't this break encoding a struct as KYAML (uses JSON field name) and then decoding it with a normal YAML decoder (ignores JSON field name)?

I'm okay with this and it makes sense in the context of Kubernetes, but it should be called out somewhere.

Reply:
Or own YAML decoders all seem happy to decode based on JSON name?  I realize now I have no idea why -- I don't see it in the yaml library.  Do you know @BenTheElder ?

Reply:
> I think our own YAML decoders work by decoding into any, re-encoding into JSON, then parsing into the struct - or at least something like that.

yeah pretty much.

https://github.com/kubernetes-sigs/yaml#introduction

https://github.com/kubernetes-sigs/yaml/blob/master/yaml.go

Reply:
So we could make the kyaml encoder respect YAML tags, inasmuch as they exist but ar non-standard, or just say it respects JSON tags because that IS standard.

Reply:
I'm fine with what your proposed (use JSON tags), but let's call out the asymmetry somewhere.

Reply:
Added a note and an UNRESOLVED to collect opinion